ssl/ssl_lib.cc

```diff
@@ -3315,3 +3315,34 @@ int SSL_CTX_set1_curves_list(SSL_CTX *ctx, const char *curves) {
 int SSL_set1_curves_list(SSL *ssl, const char *curves) {
   return SSL_set1_groups_list(ssl, curves);
 }
+
+size_t SSL_client_hello_get0_ciphers(SSL *ssl, const unsigned char **out) {
+  STACK_OF(SSL_CIPHER) *client_cipher_suites = SSL_get_client_ciphers(ssl);
+  if (!client_cipher_suites) {
+      return 0;
+  }
+
+  if (out) {
+    // Called before
+    if(!ssl->client_cipher_suites_arr.empty()) {
+      *out = reinterpret_cast<unsigned char*>(ssl->client_cipher_suites_arr.data());
+      return ssl->client_cipher_suites_arr.size();
+    }
+
+    size_t total_length = sk_SSL_CIPHER_num(client_cipher_suites);
+    assert(ssl->client_cipher_suites_arr.Init(total_length));
+
+    // Construct list of cipherIDs
+    for (size_t i = 0; i < sk_SSL_CIPHER_num(client_cipher_suites); i++) {
+      const SSL_CIPHER *cipher = sk_SSL_CIPHER_value(client_cipher_suites, i);
+      uint16_t iana_id = SSL_CIPHER_get_protocol_id(cipher);
+
+      CRYPTO_store_u16_be(&ssl->client_cipher_suites_arr[i], iana_id);
+    }
+
+    *out = reinterpret_cast<unsigned char*>(ssl->client_cipher_suites_arr.data());
```

Comment:
Is it safe to do this cast from an array of uint16_t to unsigned chars which are closer to uint8t_? Will this hit any endianness issues?

Reply:
Based on my research this should be generally safe. When casting to unsigned char *, the uint16_t is treated as an array of bytes. So the difference may just be it takes 2 bytes of memory, 2 elements of the array. Since we're using CRYPTO_store_u16_be to explicitly specify order, the endianness should be correct.